File: scripts/gesture_DTW.py
#!/usr/bin/env python

import rospy
import numpy as np

from collections import deque

from geometry_msgs.msg import Vector3
from std_msgs.msg import Int8

gesture_dict = {0: "backward",
                1: "forward",
                2: "turn_right/cw",
                3: "turn_left/ccw"}


class GestureDTW:
    def __init__(self):
        # set parameters
        self.num_gestures = rospy.get_param("~num_gestures", 4)
        self.max_window_size = rospy.get_param("~max_window_size", 45)
        self.dtw_window = rospy.get_param("~dtw_window", 5)

        self.gesture_temps = []
        self.gesture_len = []
        self.rej_thresh = []
        self.rej_corr = [1.5, 1.5, 1.5, 1.5]    # should load from file

        # obtain data file path through parameter
        data_file_path = rospy.get_param("~gesture_data_path", "../gesture_data")
        self.__load_data__(data_file_path)

        # acceleration data input
        self.acc_raw = np.array([0, 0, 0], dtype=np.float32)
        self.acc_filtered = deque(maxlen=self.max_window_size)
        self.acc_hist_len = 0

        # recognized gesture number
        self.gesture_id = Int8()

        # a flag that indicates that a gesture has been detected
        self.flag_gesture = False

        # a mini state machine for debouncing
        self.gesture_state = 0
        self.time_gesture_start = 0.0
        self.time_debounce = 0.5

        # subscriber and publisher
        self.acc_data_sub = rospy.Subscriber("human_input/acc_raw",
                                             Vector3, self.acc_data_callback)
        self.gesture_pub = rospy.Publisher("human_input/gesture",
                                           Int8, queue_size=1)

    def __load_data__(self, data_file_path):
        self.num_gestures = 4

        # open and read the threshold file
        data_file = open(data_file_path + "/rejection_threshold.txt", 'r')
        for g in range(0, self.num_gestures):
            thresh_data = data_file.readline()
            self.rej_thresh.append(float(thresh_data))

        # close the file
        data_file.close()

        # load the gesture template files
        for g in range(0, self.num_gestures):
            data_file = open(data_file_path + "/temp_gesture" + str(g) + ".txt", 'r')
            data_all = data_file.readlines()

            temp = np.zeros((len(data_all), 3), dtype=np.float32)
            temp_len = 0

            for data_point in data_all:
                temp[temp_len, :] = np.array(map(float, data_point.split(',')))
                temp_len += 1

            self.gesture_temps.append(temp)
            self.gesture_len.append(temp_len)

            # close the file
            data_file.close()

    def acc_data_callback(self, msg):
        self.acc_raw[0] = msg.x
        self.acc_raw[1] = msg.y
        self.acc_raw[2] = msg.z
        # print self.acc_raw

        # add data to acc history queue
        self.add_filtered_acc()

        # do gesture detection
        if self.gesture_state == 0:
            # do detection
            self.gesture_matching()

            if self.flag_gesture:
                # publish the gesture
                self.gesture_pub.publish(self.gesture_id)

                # reset the flag
                self.flag_gesture = False

                # switch to not detect state
                self.gesture_state = 1
                self.time_gesture_start = rospy.get_time()
        else:
            # no gesture detection, just check time
            if rospy.get_time() - self.time_gesture_start > self.time_debounce:
                self.gesture_state = 0

    def add_filtered_acc(self):
        # just a simple IIR filter
        if not self.acc_filtered:
            self.acc_filtered.append(self.acc_raw)
        else:
            acc_last = self.acc_filtered.pop()
            acc_new = 0.5 * acc_last + 0.5 * self.acc_raw

            self.acc_filtered.append(acc_last)
            self.acc_filtered.append(acc_new)

    def gesture_matching(self):
        gesture_score = []
        self.flag_gesture = False

        # only start to do gesture detection after having enough data
        if len(self.acc_filtered) < self.max_window_size:
            return

        # do DTW with each gesture template
        for g in range(0, self.num_gestures):
            W = 1e6 * np.ones((self.max_window_size+1, self.gesture_len[g]+1), dtype=np.float32)
            W[0, 0] = 0
            gesture = self.gesture_temps[g]

            for i in range(0, self.max_window_size):
                acc_i = self.acc_filtered.popleft()
                for j in range(max(0, i - self.dtw_window), min(self.gesture_len[g], i + self.dtw_window)):
                    W[i+1, j+1] = np.linalg.norm(acc_i - gesture[j, :]) + \
                                  min(W[i, j+1], W[i+1, j], W[i, j])

                self.acc_filtered.append(acc_i)

            gesture_score.append(W[self.max_window_size, self.gesture_len[g]] / self.gesture_len[g])
            if gesture_score[g] > self.rej_thresh[g] * self.rej_corr[g]:
                # gesture_score[g] = 1e6
                pass
            else:
                self.flag_gesture = True

        # print gesture_score
        # find the best match
        if self.flag_gesture:
            self.gesture_id = gesture_score.index(min(gesture_score))
            print gesture_dict[self.gesture_id]
            self.gesture_pub.publish(self.gesture_id)

if __name__ == "__main__":
    # initialize node and a gesture recognition object
    rospy.init_node("gesture_detector")
    gesture_rec = GestureDTW()

    # loop rate 50Hz
    rate = rospy.Rate(50)
    while not rospy.is_shutdown():
        rate.sleep()

File: src/gesture_DTW.cpp
#include "gesture_DTW.h"
#include <iostream>
#include <fstream>
#include <sstream>

// ============================================================================
// constructor
GestureDTW::GestureDTW()
{
    // initialize subscriber and publishers
    this->acc_data_sub = this->nh.subscribe<geometry_msgs::Vector3>("/human_input/acc_raw",
                                                                    1, &GestureDTW::acc_data_callback, this);
    this->gesture_pub = this->nh.advertise<std_msgs::Int8>("/human_input/gesture", 1);

    // initialize the "constants"
    ros::param::param<int>("~num_gestures", this->num_gestures, 6);
    ros::param::param<int>("~max_window_size", this->max_window_size, 45);
    ros::param::param<int>("~dtw_window", this->dtw_window, 45);

    this->time_debounce = 0.5;
    this->time_gesture_start = 0;

    double vals[] = {1.5, 1.5, 1.5, 1.5, 1.5, 1.5};
    this->rej_corr.assign(vals, vals+this->num_gestures);

    // flags
    this->flag_gesture = false;

    //! initialize state
    this->state = State_Detection_Start;

    // load file
    std::string file_path;
    ros::param::param<std::string>("~gesture_data_path", file_path, "gesture_data");
    this->load_data(file_path);
}

// ============================================================================
// distructor
GestureDTW::~GestureDTW()
{
}

// ============================================================================
void GestureDTW::acc_data_callback(const geometry_msgs::Vector3::ConstPtr &msg)
{
    this->acc_raw(0) = msg->x;
    this->acc_raw(1) = msg->y;
    this->acc_raw(2) = msg->z;

    // filter data and add to history
    this->add_filtered_acc();

    // state machine for gesture detection
    switch (this->state)
    {
    case State_Detection_Idle:
        break;

    case State_Detection_Start:
        // perform gesture detection
        this->gesture_matching();

        // switch to debounce once gesture detected
        if (this->flag_gesture) {
            // publish gesture
            this->gesture_pub.publish(this->gesture_id);

            this->flag_gesture = false;
            this->state = State_Detection_Debounce;
            this->time_gesture_start = ros::Time::now().toSec();
            // std::cout << "gesture " << this->gesture_id << std::endl;
        }
        break;

    case State_Detection_Debounce:
        // check time
        double t_debounce = ros::Time::now().toSec() - this->time_gesture_start;
        if (t_debounce >= this->time_debounce) {
            // go back to detection state
            this->state = State_Detection_Start;
        }
        break;
    }
}

// ============================================================================
void GestureDTW::load_data(std::string file_path)
{
    // open and read the threshold file
    std::ifstream data_file;
    std::stringstream full_path;
    full_path << file_path << "/rejection_threshold.txt";
    data_file.open(full_path.str().c_str());

    double new_thresh;
    for (int g = 0; g < this->num_gestures; g++)
    {
        data_file >> new_thresh;
        this->rej_thresh.push_back(new_thresh);
    }
    data_file.close();

    // open and read the gesture file
    for (int g = 0; g < this->num_gestures; g++)
    {
        full_path.str("");
        full_path << file_path << "/template" << g << ".txt";
        data_file.open(full_path.str().c_str());

        // get data into a vector first
        std::vector<std::string> data;
        std::string line;
        while (std::getline(data_file, line))
        {
            data.push_back(line);
        }

        // convert to matrix
        Eigen::MatrixXd temp(data.size(), 3);
        for (int i = 0; i < data.size(); i++)
        {
            std::stringstream ss(data[i]);
            ss >> temp(i, 0);
            ss.ignore(1);
            ss >> temp(i, 1);
            ss.ignore(1);
            ss >> temp(i, 2);
            std::cout << temp(i, 0) << " " << temp(i, 1) << std::endl;
        }

        // add gesture template
        this->gesture_temps.push_back(temp);
        std::cout << "template " << g << " loaded!" << std::endl;

        // close the file
        data_file.close();
    }
}

// ============================================================================
void GestureDTW::add_filtered_acc()
{
    if (this->acc_filtered.empty()) {
        this->acc_filtered.push_back(this->acc_raw);
    } else {
        Eigen::Vector3d acc_new = 0.5 * this->acc_raw + 0.5 * this->acc_filtered.back();
        this->acc_filtered.push_back(acc_new);

        // maintain the size
        if (this->acc_filtered.size() > this->max_window_size) {
            this->acc_filtered.pop_front();
        }
    }
}

// ============================================================================
void GestureDTW::gesture_matching()
{
    double gesture_score = 0;
    double min_score = 1e6;

    // only do gesture detection after having enough data
    if (this->acc_filtered.size() < this->max_window_size) {
        return;
    }

    // perform DTW on each gesture template
    for (int g = 0; g < this->num_gestures; g++) {
        Eigen::MatrixXd temp = this->gesture_temps[g];
        int temp_len = temp.rows();

        // initialize the cost array
        double w[this->max_window_size+1][temp_len+1];
        for (int i = 0; i < this->max_window_size+1; i++) {
            for (int j = 0; j < temp_len+1; j++) {
                w[i][j] = 1e6;
            }
        }
        w[0][0] = 0;

        // DTW
        for (int i = 0; i < this->max_window_size; i++) {
            w[i+1][0] = w[i][0] + dist(this->acc_filtered[i], temp.row(0));
        }

        for (int j = 0; j < temp_len; j++) {
            w[0][j+1] = w[0][j] + dist(this->acc_filtered[0], temp.row(j));
        }

        for (int i = 0; i < this->max_window_size; i++) {
            int j_start = std::max(0, i - this->dtw_window);
            int j_end = std::min(temp_len, i + this->dtw_window);
            for (int j = j_start; j < j_end; j++) {
                w[i+1][j+1] = std::min(w[i][j+1], std::min(w[i+1][j], w[i][j])) +
                        dist(this->acc_filtered[i], temp.row(j));
            }
        }

        // check if greater than threshold
        gesture_score = w[this->max_window_size][temp_len] / temp_len;
        if (gesture_score < this->rej_thresh[g] * this->rej_corr[g]){
            this->flag_gesture = true;
            if (gesture_score < min_score) {
                min_score = gesture_score;
                this->gesture_id.data = g;
            }
        }
//        std::cout << gesture_score << " ";
    }
//    std::cout << std::endl;
}

// ============================================================================
double GestureDTW::dist(Eigen::VectorXd v1, Eigen::VectorXd v2)
{
    Eigen::VectorXd v_diff = v1 - v2;
    return v_diff.norm();
}

// ============================================================================
int main(int argc, char** argv)
{
    ros::init(argc, argv, "gesture_recognition");
    GestureDTW gesture_rec = GestureDTW();

    ros::Rate loop_rate(50);
    while (!ros::isShuttingDown()) {
        ros::spinOnce();
        loop_rate.sleep();
    }
}
